proposals/052-cryptography-leg-1.md

```diff
@@ -0,0 +1,329 @@
+
+> NOTE: Cryptocurrency libraries and libraries not on Hackage are excluded from consideration.
+
+Specific libraries worth noting are:
+
+- Predecessors to `cryptonite` - all deprecated
+    - [cryptohash](https://hackage.haskell.org/package/cryptohash)
+    - [cryptocipher](https://hackage.haskell.org/package/cryptocipher)
+    - [crypto-pubkey](https://hackage.haskell.org/package/crypto-pubkey)
+    - [crypto-random](https://hackage.haskell.org/package/crypto-random)
+- Cryptographic "kitchen sinks"
+    - [cryptonite](https://hackage.haskell.org/package/cryptonite)
+    - [crypton](https://hackage.haskell.org/package/cryptonite) - a fork of cryptonite
+- Bindings to libsodium
+    - [NaCl](https://github.com/serokell/haskell-crypto)
+    - [saltine](https://hackage.haskell.org/package/saltine)
+    - [sel](https://hackage.haskell.org/package/sel-0.0.1.0/candidate)
+- Bindings to botan
+    - [Z-Botan](https://hackage.haskell.org/package/Z-Botan) - part of Z-Haskell, GHC8
+- Honorable mentions
+    - [Crypto](https://hackage.haskell.org/package/Crypto) - a very old sink
+    - [crypto-api](https://hackage.haskell.org/package/crypto-api) - cryptographic classes
+    - [cryptol](https://hackage.haskell.org/package/cryptol) - a DSL for developing new cryptographic primitives
+
+These prior efforts can be considered to be somewhat partially successful in presenting Haskell developers with access to cryptographic primitives, but present a significant development challenge and a high refactoring cost if dependences become deprecated, which happens often. Of these libraries, a considerable amount are deprecated or have been dead for many years, with `cryptonite`, `crypton`, and `saltine` being clear front-runners in terms of current usage and transitive dependencies (and `sel` being under active development). Of these four, one is a fork of another (`crypton` and `cryptonite`), and the other two are bindings to `libsodium` which does not sufficiently capture our need for low-level cryptographic primitives in general.
+
+None of these sufficiently express cryptography through the type-system and with type-classes. `crypton/ite` exposes a few highly- opinionated / specific classes for hashing, block and stream ciphers, but otherwise is focused on concrete implementations and functions. Both `crypton` and `cryptonite` rely on bindings to C implementations of unknown veracity; though the code is itself not suspect, the lack of provenance makes it ill-suited for recommendation in trusted environments, and maintenance of the C implementation poses burden of high technical skill. Libsodium bindings are more trustworthy, but the limited options severely hamper interop with other systems unless they too are using libsodium, making them ill-suited for recommendation for some use cases. `Z-Botan` has been apparently dead for several years, and even if revived has a has a sizeable dependency in the form of the `Z-Haskell` ecosystem.
```

Comment:
While many of my comments here are nitpicks, this one is substantive.

The proposal treats type classes for cryptography as self-evidently good, but I don't know what the use case for them is. It would make the text much more convincing to have at least an example of where this would make programmers' lives better with respect to just importing the function they need.

Reply:
Why are you proposing to create a new set of bindings in response to old ones being unmaintained? Why will this effort be maintained in the long run, unlike these others?

Reply:
The more abstract stuff such as typeclasses are being removed from this proposal for clarity. They've proven to be distracting, and I've narrowed the scope to just the `botan-bindings`, `botan-low`, and `botan` libraries. There are advantages to cryptography typeclasses, but their utility is more apparent when there are multiple backends (of which Botan would be the first). In particular, I believe that cryptographic typeclasses would find use in developing backend-agnostic cryptography test suites. Another issue is that hardcoding of algorithms is detrimental in the long run - it makes it difficult to patch vulnerabilities and switch to newer algorithms when attacks are discovered. A few examples of this is worldwide usage of MD5 in general, and specifically Github / SHA2 recently.

Regarding the old unmaintained bindings, `Z-Botan` is extremely tightly bound to the entire `Z-Haskell` ecosystem, and would require comprehensive and near-total rewrite in order to divest itself of `Z-Haskell` as a dependency. These new bindings draw significantly from `Z-Botan`, but present a much smaller maintenance and dependency surface.